# functions for rapidly retrieving spans of reference genome bases
# expects environment variables: 
#   env$GENOME_FASTA
#   env$SHM_DIR_WRK
#   env$CHROM_FASTA_DIR

#=====================================================================================
# load a genome index previously generated by samtools faidx
#-------------------------------------------------------------------------------------
loadFaidx <- function(){
    index_file <- paste(env$GENOME_FASTA, 'fai', sep = ".")
    genome_index <<- read.table(
        index_file, 
        header = FALSE, 
        sep = "\t", 
        stringsAsFactors = FALSE,
        col.names  = c('chrom',     'length',  'offset',  'lineSeqLen', 'lineLen'), 
        colClasses = c('character', 'integer', 'numeric', 'integer',    'integer')
    ) 
    genome_index$chromFile <<- sapply(genome_index$chrom, function(chrom){
        chromFile <- file.path(env$SHM_DIR_WRK, paste(chrom, "txt", sep = "."))
        if(!file.exists(chromFile)){
            fastaFile <- file.path(env$CHROM_FASTA_DIR, paste(chrom, "fa", sep = "."))
            system2("cat", c(fastaFile, "| grep -v '>' | tr -d '\n' > ", chromFile))
        } 
        chromFile       
    })
}
#=====================================================================================

#=====================================================================================
# get a specified chunk of a single chromosome and rc if requested
#-------------------------------------------------------------------------------------
getRefSeq <- function(chrom, range, rc = FALSE) {
    index <- genome_index[genome_index$chrom == chrom, , drop = FALSE]    
    if(range[1] > range[2]) range <- rev(range)
    if(range[1] < 1) range[1] <- 1L
    if(range[2] > index$length) range[2] <- index$length
    chrom_con <- file(index$chromFile, open = "rb")
    seek(chrom_con, where = range[1] - 1L, origin = "start", rw = "r")
    seq <- readChar(chrom_con, diff(range) + 1L)
    close(chrom_con)
    seq <- toupper(seq)
    if(rc) seq <- rc(seq)
    seq
}
getRefSeq_padded <- function(chrom, pos, padding, rc = FALSE){
    getRefSeq(chrom, c(pos - padding, pos + padding), rc) # pos is central base of odd-length string
}
#=====================================================================================

# #=====================================================================================
# # load a genome index previously generated by samtools faidx
# #-------------------------------------------------------------------------------------
# loadFaidx <- function(dir, genome = NULL){
#     if(is.null(genome)) genome <- env$GENOME
#     genome_file <- paste(genome, 'fa', sep = ".")
#     genome_file <- paste(dir, genome_file, sep = "/")
#     genome_index_file <- paste(genome_file, 'fai', sep = ".")
#     genome_con <<- file(genome_file, open = "rb")
#     genome_index <<- read.table(
#         genome_index_file, 
#         header = FALSE, 
#         sep = "\t", 
#         stringsAsFactors = FALSE,
#         col.names  = c('chrom',     'length',  'offset',  'lineSeqLen', 'lineLen'), 
#         colClasses = c('character', 'integer', 'numeric', 'integer',    'integer')
#         # offset too big to store as integer: .Machine$integer.max
#     ) 
# }
# #=====================================================================================
# #=====================================================================================
# # get a specified chunk of a single chromosome and rc if requested
# #-------------------------------------------------------------------------------------
# getRefSeq <- function(chrom, range, rc = FALSE) {
#     index <- genome_index[genome_index$chrom == chrom, , drop = FALSE]    
#     if(range[1] > range[2]) range <- rev(range)
#     if(range[1] < 1) range[1] <- 1L
#     if(range[2] > index$length) range[2] <- index$length
#     range[1] <- range[1] - 1L
#     range[2] <- range[2] - 1L
#     linesBeforePos1 <- as.integer(floor(range[1] / index$lineSeqLen))
#     linesBeforePos2 <- as.integer(floor(range[2] / index$lineSeqLen))
#     seek(genome_con, where = index$offset + range[1] + linesBeforePos1, origin = "start", rw = "r") # accounts for newlines # nolint
#     seq <- readChar(genome_con, diff(range) + 1L + (linesBeforePos2 - linesBeforePos1))
#     seq <- toupper(gsub("[[:space:]]", "", seq))
#     if(rc) seq <- rc(seq)
#     seq
# }
# getRefSeq_padded <- function(chrom, pos, padding, rc = FALSE){
#     getRefSeq(chrom, c(pos - padding, pos + padding), rc) # pos is central of odd-length string
# }
# #=====================================================================================
